design/ceph/cephx-keyring-rotation.md

```diff
@@ -0,0 +1,483 @@
+
+### Enabling CSI key rotation
+
+Ceph-CSI deployments (if managed by Rook) are managed in the operator namespace, but keys are
+created on a per-CephCluster basis. Thus, a CephCluster configuration option (like above) is most
+appropriate. CSI key rotations may require manual administrator action to reboot or drain/undrain
+nodes to PVCs, so Rook must allow users to initiate one-time key rotations, so CSI will use
+overlapping rotation.
+
+```yaml
+spec:
+  # ...
+  security:
+    cephx:
+      daemon: {} # discussed above
+      csi:
+        keyRotationPolicy: Disabled | KeyGeneration
+        keyGeneration: <int> # used with KeyGeneration
+        keepPriorKeyCount: 1
+      # (room for future spec.security.cephx options)
+    # (room for future spec.security options)
+  # ...
+```
+
+Note on `keyRotationPolicy` for CSI. `WithCephVersionUpgrade` will not be supported for CSI keys
+unless we can validate that the keys can safely be rotated without the risk of affecting existing
+PVC mount connectivity. Rook will return an error if this value is given.
+
+Rook could use a string like `once` as an input, but Rook would have to record when `once` was first
+observed so that it doesn't repeat rotations with every reconcile.
+
+Rook could alternatively use a string like `always` and expect the user to unset `always` after they
+see that rotation is complete, but this requires the user to monitor the rotation status and change
+config in a clunky way that might risk manual errors.
+
+Rook could use the Ceph version as the "older-than" selection, but that would only allow CSI key
+rotations when the Ceph version changes. For a user already at the latest Ceph version, they
+wouldn't have the option to rotate CSI keys on demand.
+
+Rook could use a `DATETIME` string to initiate rotation for keys minted/rotated before a certain
+time, but it is hard to track rotation time well when many keys are involved. Generation allows for
+a more clear representation of actual state and provides a simple interface.
+
+### Enabling non-daemon key rotation
+
+Ceph daemons that have keys used by non-Rook-controlled clients are also associated with
+Custom Resources (CRs).
+
+Full list of Rook CRs with non-daemon keys:
+
+- CephClient - the created arbitrarily-usable client has a single associated key
+- CephRBDMirror - mirror daemon has a mirror peer key (`peer`)
+- CephFilesystemMirror - mirror daemon has a mirror peer key (`peer`)
+
+Each CR will provide a key rotation mechanism as part of the primary API spec.
+
+```yaml
+kind: CephRBDMirror | CephFilesystemMirror
+# ...
+spec:
+  # ...
+  security:
+    cephx:
+      peer: # this distinction is not necessary but makes clear that it doesn't control daemon keys
+        keyRotationPolicy: Disabled | WithCephVersionUpgrade | KeyGeneration
+        keyGeneration: <int> # used with KeyGeneration
+    # (room for future spec.security options)
+  # ...
+```
+
+## Status reporting
```

Comment:
I can understand status is specific to specific policy used for rotation,
On high level can also add

```
RotationCompleted: True
           KeyGeneration: 
           ....
```
And other can be sub fields?